### PR TITLE
Finished All my_enumerable functions

### PR DESCRIPTION
**Pull request by AymanJabr**
In this project I re-made some of the most used Ruby Enumerable classes:

1. my_each for each
2. my_select for select
3. my_all? for all?
4. my_any? for any?
5. my_none? for none?
6. my_count for count
7. my_count() for count()
8. my_map for map
9. my_inject for inject

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED 👎🏽** 

Hey Team, Well done with the efforts towards this project. The enumerable methods pass some tests and are working almost just fine.

REQUIRED CHANGES:

- [ ] `my_each_with_index` is not working as the original `each_with_index` method, also is not passing all test cases.
- `my_each_with_index` should return an Enumerator if no block is given.

- [x] `my_select` should return an array containing all elements of enum for which the given block returns a true value. It returns `nil` instead of `[1,3,5,....]`
- [x] `my_select` returns an array containing all elements of enum for which the given block returns a true value::range. It returns `nil' instead of `[ ]`.
- [x] `my_select` should return an Enumerator if no block is given.

- [x] `my_all?` should return true if the block never returns false or nil
```
 expected: true
            got: nil

```

- [x] `my_all?` returns false if the block returns false or nil
```
 expected: false
            got: nil

```

- [x] `my_all?` should behave just like `all?`. It should be able to accept arguments passed to it.
- `my_all?` is currently throwing ArgumentError when I pass in arguments to the method.
- Happens to all integers, Numerics, regex, class passed as arguments. `Given 1, expected 0`

- [x] `my_any?` should return true if the block never returns false or nil
```
 expected: true
            got: nil

```

- [x] `my_none?` returns true if the block never returns true for all elements. `expected true, got nil`
- [x] `my_none?` returns false if the block ever returns true for all elements. `expected false, got nil`

- [x] `my_none?` should behave just like `none?`. It should be able to accept arguments passed to it.
- `my_none?` is currently throwing ArgumentError when I pass in arguments to the method.
- Happens to all integers, Numerics, regex, class passed as arguments. `Given 1, expected 0`

- [x] `my_count` returns the number of items in enum through enumeration.
- `expected 100, got nil`

- [x] `my_count` counts the number of items in enum that are equal to item if an argument is given. `returns nil`


- [x] `my_inject` should behave just like `inject`. It should be able to accept arguments passed to it.
- `my_inject` is currently throwing ArgumentError when I pass in arguments to the method.
- Happens to all integers, Numerics, regex, class passed as arguments. `Given 1, expected 0`


Apparently, your methods don't meet most of the requirements, such as yield with a block or with no blocks given. Also, yield with block and arguments given. Kindly adjust the methods I've highlighted above.

**Submit another CRR after making the changes. 💯** 

### 🔢  Happy Coding 🔢 

Reach me on slack: **Roy Mukuye**

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Required Changes 🛑♻**

<hr>

Hello there, good job you've done so far on this project. However, there are still some issues you need to fix before moving on.

- [x] Please make sure that the my_all?, my_any?, and my_none? methods are accepting arguments in types Classes like (Integer, Numeric, etc) and accepting when No Block Is Given. Kindly fix them
- [x] Please fix your my_count method. It should print out 3 when running this line of code ```puts [1,2,4,2].my_count { |x| x%2 == 0 }``` instead, it print out 4!

<hr>

Kindly fix the above changes and test your methods using given examples or examples of your own, and re-submit for another review 😊🙏

<hr>

Good luck and Happy coding 👨‍💻💻

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Ayman Jabr** 
Good job with the improvements for this project :muscle:  :muscle:
There are some improvements to do before you can go ahead:

- [x] #my_each_with_index when a block is given when self is an array
     ```
      concat1=''
      [1,2,3,4].each_with_index(&proc{|n, i| concat1+='n: '+n.to_s+', i: '+i.to_s+'+++'})
      concat2=''
      [1,2,3,4].my_each_with_index(&proc{|n, i| concat2+='n: '+n.to_s+', i: '+i.to_s+'+++'})
      concat1 == concat2
     ```
       behaves like #each_with_index calls the given block once for each element in self
- [x] #my_all? returns false if the block returns false or nil::range
    `(1..3).all?(&proc{|x| x%2==0}) == (1..3).my_all?(&proc{|x| x%2==0})`
- [x] #my_all? when no block or argument is given returns true when none of the collection members are false or nil
      `[true, [false]].all? == [true, [false]].my_all?`
- [x] #my_all? when no block or argument is given returns false when one of the collection members are false or nil
      `[true,[true],false].all? == [true,[true],false].my_all?`
- [x] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
      `['car', 'cat'].all?(/a/) == ['car', 'cat'].my_all?(/a/)`
- [x] #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
      `['car', 'cat'].all?(/t/) == ['car', 'cat'].my_all?(/t/)`
- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
      `[5,5,5].all?(5) == [5,5,5].my_all?(5)`
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any
of the collection does not match the pattern
      `[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`
- [x] #my_any? returns false if the block ever returns a value false or nil::range
    `(1..3).any?(&proc{|x| x==0}) == (1..3).my_any?(&proc{|x| x==0})`
- [x] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
      `[0,[]].all? == [0,[]].my_all?`
- [x] #my_any? when no block or argument is given returns false if at least one of the collection is not true
      `[false, 0].any? == [false, 0].my_any?`
- [x] #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
      `['dog', 'cat'].any?(/d/) == ['dog', 'cat'].my_any?(/d/)`
- [x] #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
      `['dog', 'cat'].any?(/z/) == ['dog', 'cat'].my_any?(/z/)`
- [x] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern       
      `['dog','car'].any?('cat') == ['dog','car'].my_any?('cat')`
- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
      `['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')`
- [x] #my_none? returns false if the block ever returns true for all elements::range
    `(1..3).none?(&proc{|num| num%2==0}) == (1..3).my_none?(&proc{|num| num%2==0})`
- [x] #my_none? when no block or argument is given returns true only if none of the collection members is true
      `[false, nil,false].none? == [false, nil,false].my_none?`
- [x] #my_none? when no block or argument is given returns false only if one of the collection members is true
      `[false, nil,[]].none? == [false, nil,[]].my_none?`
- [x] #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
      `['dog','cat'].none?(/x/) ==  ['dog','cat'].my_none?(/x/)`
- [x] #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
      `['dog', 'cat'].none?(/d/) == ['dog', 'cat'].my_none?(/d/)`
- [x] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern  
      `['dog','car'].none?(5) == ['dog','car'].my_none?(5)`
- [x] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern  
      `[5,'dog','car'].none?(5) == [5,'dog','car'].my_none?(5)`
- [x] #my_count returns the number of items in enum through enumeration
    `[1,2,3,4,5].count == [1,2,3,4,5].my_count`
- [x] #my_count returns the number of items in enum through enumeration::range
    `(1..3).count == (1..3).my_count`
- [x] #my_map returns an Enumerator if no block is given
    `[1,2,3].map.class==[1,2,3].my_map.class`
- [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
      `[1,2,3].inject(&proc{|total, num| total*num}) == [1,2,3].my_inject(&proc{|total, num| total*num})`
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
      `[1,2,3].inject(:+) == [1,2,3].my_inject(:+)`
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
      `(1..9).inject(:+) == (1..9).my_inject(:+)`
- [x] #my_inject when a symbol is specified with an initail value combines each element
of the collection by applying the symbol as a named method
      `[1,2,3].inject(4, :*) == [1,2,3].my_inject(4, :*)`
- [x] #my_inject when a symbol is specified with an initail value combines each element
of the collection by applying the symbol as a named method::range
      `(1..3).inject(4, :*) == (1..3).my_inject(4, :*)`
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
    `multiply_els([1,2,3]) == 6`
- [x] Add a screenshot in the README file

Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
### Hey there :wave: 
##### You have done well so far and just a little bit more for the project to be approved. You have a few failing tests that you need to work on.
- [ ] enumerables #my_each_with_index when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
     
       expected: "Num: 6, idx: 0\nNum: 8, idx: 1\nNum: 2, idx: 2\nNum: 1, idx: 3\nNum: 4, idx: 4\nNum: 8, idx: 5\nNum:...: 8, idx: 94\nNum: 0, idx: 95\nNum: 1, idx: 96\nNum: 0, idx: 97\nNum: 3, idx: 98\nNum: 0, idx: 99\n"
            got: "Num: 6, idx: 0\nNum: 8, idx: 1\nNum: 2, idx: 2\nNum: 1, idx: 3\nNum: 4, idx: 4\nNum: 8, idx: 1\nNum:...um: 8, idx: 1\nNum: 0, idx: 11\nNum: 1, idx: 3\nNum: 0, idx: 11\nNum: 3, idx: 12\nNum: 0, idx: 11\n"
- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
     Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)
     
       expected true
            got false
- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
     Failure/Error: expect(numbers.my_all?(Numeric)).to be numbers.all?(Numeric)
     
       expected true
            got false
- [x] enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
     Failure/Error: expect(array.my_all?(3)).to be array.all?(3)
     
       expected false
            got true
- [x] enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
     Failure/Error: expect(array.my_any?(Numeric)).to be array.any?(Numeric)
     
       expected true
            got false
- [x] enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
     Failure/Error: expect(array.my_inject(&operation)).to eq array.inject(&operation)
     
       expected: 412
            got: 379
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
     Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)
     
       expected: 420
            got: 316
- [x] enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
     Failure/Error: actual = multiply_els [2, 4, 5]
##### do request another review when you have made this changes
slack(@ebeagusamuel)

File: .rubocop.yml
Comment (BrianSammit) [inline .rubocop.yml]: In line 11, you change the max:20 to max:50, keep the original file. You can disable a particular rubocop. 
In this case use: 
Put this line at the beginning  of the code 
`# rubocop:disable Metrics/MethodLength`
Put this line at the end of the code 
`# rubocop:enable Metrics/MethodLength`

File: index.rb
Comment (BrianSammit) [inline index.rb]: From line 224 to line 235 are debugging code. kindly delete or keep for yourself on another branch. :hammer_and_wrench:

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :recycle: :red_circle: 
---------------------
Hello team :wave: you are working really hard on this project congrats on that :muscle: :1st_place_medal: 
There are a few changes to make :hammer_and_wrench: see them below :point_down: 
------------

- [ ] You change the `rubocop.yml`. Do not make any changes in config files - they represent style guidelines that you share with your team - which is a group of all Microverse students. Check my inline comment to know how to deal with this. 
- [x] Delete or keep on another branch the debugging code. [Here](https://questions.microverse.org/t/debugging-instructions-code/346) you can read about why. Check my inline comments to know what is the code I'm talking about. 
- [x] #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class. Instaed is returning True. 
- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern. Instead is returning False
- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer. Instead is returning True.
- [x] #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block. 
- [x] #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method.
-------------------
Make the required changes and send it for another review :+1: :muscle: 
HAPPY CODDING !!!

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @AymanJabr 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

Kindly work on the suggested changes 👇.
- [ ] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- [ ] `#my_inject `when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- [ ] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- [ ] `#multiply_els` accepts an array as an argument and multiplies all the elements of the array together using `#my_inject`
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

File: index.rb
Comment (ebeagusamuel) [inline index.rb]: - [x]  There is a failing test on your `my_each_with_index` It doesn't return the correct index and value when tested with the original `each_with_index` 
P.S You are not meant to use the original enumerable methods in your customer method. The idea was for you to create your own method that behaves like the original one and not to use the original method.

File: index.rb
Comment (ebeagusamuel) [inline index.rb:90]: - [x] There is also an error in your implementation of the my_any?. `%w[dog door rod blade].any?(Integer)` returns `false` while `%w[dog door rod blade].my_any?(Integer)` return `true`

File: index.rb
Comment (ebeagusamuel) [inline index.rb]: - [x] This should not be inside the enumerable block. It should be a standalone method.

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
### Hey there :wave: 
##### There are still changes required. I have tried to explain those changes in the comments below. Please do check them and make changes accordingly.
slack(@ebeagusamuel)

File: index.rb
Comment (mahmoud717) [inline index.rb]: - [x] my_each should return an Enumerator if no block is given.

File: index.rb
Comment (mahmoud717) [inline index.rb]: - [x] my_each_with_index should return an Enumerator if no block is given.
- [x] my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self. that means you have to return the original array after operating on it.
- [x] my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self. that means it should be able to accept ranges.
- [x] my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self. that means it should accept hashes and operate correctly on them.

File: index.rb
Comment (mahmoud717) [inline index.rb:91]: - [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric. that means it should be able to do this 
([1, "a" ].my_any(Numeric)  == true ) # true

File: index.rb
Comment (mahmoud717) [inline index.rb]: - [x] my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block. that means it has to accept blocks with operation like { |sum, n| sum + n }  and return the correct answer.
- [x] my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method. that means it has to accept a symbol like this [1,2].my_inject(:+). and this should combine all elements.

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello there**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] check my inline comments.

---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

File: index.rb
Comment (AymanJabr) [reply index.rb]: My inject works with both of the test cases mentioned above, but reviewers are saying it doesn't, I don't get why that is.

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED 🔴** 

### Hey Team,
**Thanks for all your efforts towards getting the project approved. We have just three more issues to fix.**

### REQUIRED CHANGES: 🔴 

- [ ] `my_each_with_index` when a block is given when self is an array behaves like `each_with_index` that calls the given block once for each element in self.
- [ ] ` my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block. For a given block, while `expecting 406` but `got 278`.
- [ ] ` my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method. for a ` + ` specified, `expecting 435` by `got 323` instead.

### Kindly make those three fixes and submit another CRR. 😄 

### Cheers! 🍹 

Reach me on slack for a chit chat 🆘  **Roy Mukuye**

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello there**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] please fix all of the issues requested by the previous tse and request another code review. 
- [ ] if you are having trouble with the requirements you can use [this ](https://github.com/gurbuzertunga/ruby_enumerables)project as a reference.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

File: index.rb
Comment (enaburekhan) [inline index.rb]: #my_inject raises a "LocalJumpError" when no block or argument is given.
Expected localJumpError. instead, got NoMethodError: undefined method  'yield'. You need to refactor the code to correctly use yield.
Using the regular built-in enumerable method like **each** is not allowed. We are supposed to create methods that mimic them.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Request changes:full_moon:
**Great work so far on this milestone:clap::clap:. However, we need to make a few corrections before we can continue and get this project approved:muscle:.**

- [ ] should have working `my_each` **(covered by automated tests)** identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
We can then reuse the custom my_each throughout your codebase and not the regular built-in **each** method

Make the required changes and when you are ready, submit for another review.
Kindly drop your questions:point_down: and ping me on slack @ Eric Enaburekhan to respond.

Happy Coding:blush:
Eric Enaburekhan

**Review by acushlakoncept — APPROVED** [review]
# Status: PROJECT APPROVED  🎉 🚀

### Congratulations Ayman on completing your project and taking the time to implement the changes requested by the TSEs. 

Your enumerable methods now pass all the tests. 👏 👏 👏 

I will go ahead and mark this as complete, feel free to merge when you are ready. All the best with the next project.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Pull request merged**
